controllers/druid/drain_historical.go

```diff
@@ -0,0 +1,191 @@
+		items := listObj.(*v1.PodList).Items
+		result := make([]object, len(items))
+		for i := 0; i < len(items); i++ {
+			result[i] = &items[i]
+		}
+		return result
+	})
+	if err != nil {
+		return err
+	}
+
+	pvcLabels := map[string]string{
+		"druid_cr":          m.Name,
+		"nodeSpecUniqueStr": nodeSpecUniqueStr,
+		"component":         nodeSpec.NodeType,
+	}
+
+	pvcList, err := readers.List(context.TODO(), sdk, m, pvcLabels, emitEvent, func() objectList { return &v1.PersistentVolumeClaimList{} }, func(listObj runtime.Object) []object {
+		items := listObj.(*v1.PersistentVolumeClaimList).Items
+		result := make([]object, len(items))
+		for i := 0; i < len(items); i++ {
+			result[i] = &items[i]
+		}
+		return result
+	})
+	if err != nil {
+		return err
+	}
+
+	podNames := getPodNames(podList)
+	//start draining the historical pod in batches and delete them
+	batchCount := originalReplicaCount/nodeSpec.DeploymentConfig.BatchSize + 1
+
+	//TODO add auth
+	d := InitClient(baseUrl, userName, os.Getenv("DRUID_PASSWORD"))
+	CoordinatorConfig, err := d.GetCoordinatorConfig()
+
+	if err != nil {
+		return err
+	}
+	startPod := m.Status.Historical.CurrentBatch * nodeSpec.DeploymentConfig.BatchSize
+	endPod := startPod + nodeSpec.DeploymentConfig.BatchSize - 1
+	if m.Status.Historical.CurrentBatch == batchCount {
+		endPod = originalReplicaCount - 1
+	}
+	if m.Status.Historical.DecommissionedPods == nil {
+		for podId := startPod; podId <= endPod; podId++ {
+			m.Status.Historical.DecommissionedPods = append(m.Status.Historical.DecommissionedPods, podNames[podId])
+		}
+		CoordinatorConfig["decommissioningNodes"] = []string(m.Status.Historical.DecommissionedPods) //TODO fix this
+		err := d.UpdateCoordinatorConfig(CoordinatorConfig)
+		if err != nil {
+			return err
+		}
+	}
+	//wait for draining to complete so we can start with the next batch
+	drainStatus, err := checkDrainStatus(d, startPod, endPod, podNames)
+	if err != nil {
+		return err
+	}
+	if !drainStatus {
+		logger_drain_historical.Info("Waiting for pods to drain", "name", m.Name, "namespace", m.Namespace)
+		return nil
+	}
+	//delete corresponding nodegrabbers
+	for m.Status.Historical.DecommissionedPods != nil {
+		//get the PVC for the pod to be disabled and delete it once the pod is deleted
+		for _, pod := range podList {
+			if pod.(*v1.Pod).Name == m.Status.Historical.DecommissionedPods[0] {
+				err := writers.Delete(context.TODO(), sdk, m, pod, emitEvent, &client.DeleteOptions{})
+				if err != nil {
+					return err
+				} else {
+					msg := fmt.Sprintf("Deleted pod [%s] in namespace [%s], since it was drained completely", pod.(*v1.Pod).Name, pod.(*v1.Pod).Namespace)
+					logger_drain_historical.Info(msg, "Object", stringifyForLogging(pod.(*v1.Pod), m), "name", m.Name, "namespace", m.Namespace)
+					m.Status.Historical.DecommissionedPods = m.Status.Historical.DecommissionedPods[1:]
+				}
+			}
+		}
+	}
+	for it := startPod; it <= endPod; it++ {
```

Comment:
After you delete the PVCs, the new pod which was deleted just before this block, would see the reference to the PVC and report that the PVC is missing and hence can't schedule. This is why we may need an additional round of deletion of the pods we deleted earlier.